app/controllers/authenticate_controller.rb

```diff
@@ -1,30 +1,89 @@
 class AuthenticateController < ApplicationController
   include TokenGenerator
 
-  # Request path indicates the username for +authenticate+.
-  before_filter :credentials_lookup
+  class MemoizedRole
+    def self.[](role_id)
+      @user_roles ||= Hash.new { |h, id| h[id] = Role[id] }
+      @user_roles[role_id]
+    end
+
+    def self.roleid_from_username(account, username)
+      Role.roleid_from_username(account, username)
+    end
+  end
 
   def authenticate
-    api_key = request.body.read
-    if @credentials.valid_api_key? api_key
-      render json: authentication_token
+
+    authenticator  = params[:authenticator]
+    service_id     = params[:service_id]
+    account        = params[:account]
+    username       = params[:id]
+    password       = request.body.read
+    role_id        = MemoizedRole.roleid_from_username(account, username)
+
+    validate_role_exists!(role_id)
+    
+    case authenticator
+    when 'authn-ldap'
+      validate_security!(authenticator, account, service_id, username)
+      raise Unauthorized unless ldap_authenticator.valid?(username, password)
+      role = MemoizedRole[role_id]
+    # default conjur: authn
     else
```

Comment:
we need to be careful that we send a 401 when the `authenticator` is something besides a custom authn that is handled explicitly or exactly `authn`. that is, we shouldn't have a blanket `else` here - it should be `when 'authn'` (do the usual standard authn stuff) and `else` 401.